R/register_extendr.R

```diff
@@ -40,27 +45,38 @@ register_extendr <- function(path = ".", quiet = FALSE, force_wrappers = FALSE,
 
   outfile <- rprojroot::find_package_root_file("R", "extendr-wrappers.R", path = path)
 
-  # If force_wrappers is TRUE, use a minimal wrapper file even when
-  # make_wrappers() fails; since the wrapper generation depends on the compiled
-  # Rust code, the package needs to be installed before attempting this, but
-  # it's not always the case (e.g. the package might be corrupted, or not
-  # installed yet).
-  if (isTRUE(force_wrappers)) {
-    error_handle <- function(e) {
-      cli::cli_alert_danger("Failed to generate wrapper functions: {e$message}.")
-      cli::cli_alert_warning("Falling back to a minimal wrapper file instead.")
-
-      make_example_wrappers(pkg_name, outfile, path = path)
-    }
-  } else {
-    error_handle <- function(e) {
-      ui_throw(
-        "Failed to generate wrapper functions.",
-        c(
-          ui_x(e[["message"]])
-        )
-      )
-    }
+  path <- rprojroot::find_package_root_file(path = path)
+
+  # If compile is NA, compile if the DLL is newer than the source files
+  if (isTRUE(is.na(compile))) {
+    compile <- needs_compilation(path, quiet) || pkgbuild::needs_compile(path)
+  }
+
+  if (isTRUE(compile)) {
+    # This relies on [`pkgbuild::needs_compile()`], which
+    # does not know about Rust files modifications.
+    # `force = TRUE` enforces compilation.
+    pkgbuild::compile_dll(
+      path = path,
+      force = TRUE,
+      quiet = quiet
+    )
+  }
+
+  dll_file <- get_library_path(path)
+
+  # If the wrapper file is newer than the DLL file, assume it's generated by
+  # the latest DLL, which should mean it doesn't need regenerated. This isn't
+  # always the case (e.g. when the user accidentally edited the wrapper file
+  # by hand) so the user might need to run with `force = TRUE`.
+  if (!isTRUE(force) &&
+      isTRUE(file.exists(dll_file)) &&
+      isTRUE(file.exists(outfile)) &&
+      isTRUE(file.info(dll_file)$mtime < file.info(outfile)$mtime)) {
+    cli::cli_alert_info("{.file {outfile}} is up-to-date. Skip generating wrapper functions.")
```

Comment:
I think it is better to have `{.file {pretty_rel_path(outfile)}}` than just `outfile`. {rprojroot} should produce an absolute path, so `outfile` is very long. `{pretty_rel_path(outfile)}` outputs path relative to the package root, so something like `R/extendr-wrappers.R`.